src/main/kotlin/de/smartsquare/starter/mqtt/MqttHandler.kt

```diff
@@ -18,57 +23,76 @@ class MqttHandler(
 ) {
 
     private val logger = LoggerFactory.getLogger(this::class.java)
+    private val subscriberCache = ConcurrentHashMap<MqttTopic, MqttSubscriber>(collector.subscribers.size)
+    private val lookup = MethodHandles.publicLookup()
 
-    private val subscriberCache = ConcurrentHashMap<MqttTopic, ResolvedMqttSubscriber>(collector.subscribers.size)
+    private data class MqttSubscriber(val subscriber: AnnotatedMethodDelegate, val parameterTypes: List<Class<*>>)
+
+    private fun interface AnnotatedMethodDelegate {
+        fun invoke(vararg args: Any)
+    }
 
     /**
      * Handles a single [message]. The topic of the message is used to determine the correct subscriber which is then
      * invoked with parameters produced by the [MqttMessageAdapter].
      */
     fun handle(message: MqttPublishContainer) {
-        val topic = message.topic
+        val (topic, payload) = message
+        if (logger.isTraceEnabled) logger.trace("Received mqtt message on topic [$topic] with payload $payload")
+        val (subscriber, parameterTypes) = getSubscriber(topic)
+        val adaptedParameterResult = adaptParameters(parameterTypes, message, topic)
 
-        if (logger.isTraceEnabled) {
-            logger.trace("Received mqtt message on topic [$topic] with payload ${message.payload}")
-        }
+        adaptedParameterResult
+            .mapCatching { args -> subscriber.invoke(*args) }
+            .onFailure { t ->
+                val mqttMessageException = t as? MqttMessageException
+                    ?: MqttMessageException(topic, payload, "Error while handling mqtt message on topic [$topic]", t)
 
-        val subscriber = subscriberCache
-            .getOrPut(topic) { collector.subscribers.find { it.topic.matches(topic) } }
+                messageErrorHandler.handle(mqttMessageException)
+            }
+    }
+
+    /**
+     * Returns the subscriber for the given [topic].
+     * If no subscriber is found, an error is thrown.
+     * The subscriber is cached for performance reasons.
+     *
+     * If the function is a suspend function, it is wrapped in a suspend call. For normal functions, a method handle is
+     * created and cached.
+     */
+    private fun getSubscriber(topic: MqttTopic): MqttSubscriber = subscriberCache.getOrPut(topic) {
+        val subscriber = collector.subscribers.find { it.topic.matches(topic) }
             ?: error("No subscriber found for topic $topic")
+        val kFunction = subscriber.method.kotlinFunction
+        val parameterTypes = kFunction?.valueParameters?.map { it.type.jvmErasure.java }
+            ?: subscriber.method.parameterTypes.toList()
+        val delegate = if (kFunction?.isSuspend == true) {
+            AnnotatedMethodDelegate { args -> runBlocking { kFunction.callSuspend(subscriber.bean, *args) } }
```

Comment:
As a user if you want to run your coroutine, you would normally defer to use it in the suspend function itself, as coroutines have to be written `main safe`:

```kotlin
suspend fun showCase() {
    withContext(Dispatchers.IO) {
        println("Hello, World!")
    }
}
```

There is another reason I used runBlocking here. It's about semantics of the callback processor. When using runBlocking, the thread from the callback runner ThreadPool which is running this, is actually running the code inside the runBlocking (regardless of the poor name), it's not simply blocked and waiting. That means when the suspending call is done, the thread is done running. I prefer this semantic. What's also possible though, making the executor pool schedule the suspending call in the background onto another dispatcher, for instance `Default`. This would result in the callback to just schedule the call in another dispatcher. That means the callback execution might be done, before the function inside is actually run. It would be just like submitting a task onto another ThreadPool. 

I think its more intuitive to go with the first option, thats why I picked it

Reply:
Alright, then this should be fine :+1: